File: 003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingChars.java
class LongestSubstringWithoutRepeatingChars{
    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1 ) {
            return s.length();
        }
        
        int max = 0;
        int i = 0;
        int j = 0;
        Set<Character> set = new HashSet<>();
        
        while (j < s.length() && (i < s.length())) {
            Character hi = s.charAt(i);
            Character lo = s.charAt(j);
            if (set.contains(lo)) { 
                set.remove(hi);
                i++;             
            }
            else {
                set.add(lo);
                j++;
            } 
            max = Math.max(max, j - i);
            }   
        return max;       
    }
}

File: 3 Longest Substring Without Repeating Characters/longest_substring_without_repeating_chars.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        n = len(s)
        i = j = result = 0
        while (i < n and j < n):
            if (s[j] not in char_set) :
                char_set.add(s[j])
                result = max(result, j - i + 1)
                j += 1
            else:
                char_set.remove(s[i])
                i += 1
        return result